src/main/java/uk/gov/ons/ctp/response/notify/util/BitCalculator.java

```diff
@@ -0,0 +1,207 @@
+                    "inputString='"
+                    +
+                    inputString
+                    +
+                    '\''
+                    +
+                    ", value="
+                    +
+                    value
+                    +
+                    ", valueStr='"
+                    +
+                    valueStr
+                    +
+                    '\''
+                    +
+                    ", keyName='"
+                    +
+                    keyName
+                    +
+                    '\''
+                    +
+                    ", bits="
+                    +
+                    bits
+                    +
+                    ", radix="
+                    +
+                    radix
+                    +
+                    '}';
+        }
+    }
+
+    /**
+     * Parses a String into a KeyInfo structure.  If the isKey flag is set the String is a gov.notify key and the text
+     * up to the first hyphen is assumed to be the name of the key.  If the isKey flag is not set, the string is a
+     * number in base radix with optional hyphens to break it up.
+     * @param str string to parse
+     * @param isKey key flag to indicate whether the string is a gov.notify key
+     * @param radix the radix of the number portion
+     * @return a KeyInfo object showing the number of bits of data contained within the number portion of the supplied string
+     */
+    public KeyInfo analyse(final String str, final boolean isKey, final int radix) {
+        KeyInfo info = cleanString(str, isKey, radix);
+
+        if (info.valid) {
+            int numDigits = info.valueStr.length();
+            int bitsPerDigit = new Double(getBitsPerDigit(radix)).intValue();
+            int bits = ((numDigits * bitsPerDigit) >>> radix);
+            boolean valid = isKey == false || bits >= MINIMUM_KEY_BITS;
+
+            return new KeyInfo(info.inputString, info.value, info.valueStr, info.keyName, bits, info.radix, valid);
+        } else {
+            return info;
+        }
+    }
+
+    /**
+     * Analyse a number encoded as a hexadecimal string
+     * @param hexString A number encoded as a hex string
+     * @return an object showing the number of bits of data in the supplied string
+     */
+    public KeyInfo analyseHexNumber(final String hexString) {
+        return analyse(hexString, false, HEX_RADIX);
+
+    }
+
+    /**
+     * Analyse a potential UUID. If it's a valid hex number and there are exactly 128 bits, the returned structure
+     * will be have valid = true (otherwise false)
+     * @param hexString A number encoded as a hex string
+     * @return an object showing the number of bits of data in the supplied string and a flag showing whether it's a valid #
+     * UUID
+     */
+    public KeyInfo analyseUUID(final String hexString) {
+        final KeyInfo info = analyse(hexString, false, HEX_RADIX);
+
+        if (info.bits != UUID_BITS && info.valid == true) {
```

Comment:
unnecessary comparison to true

Reply:
Stylistic thing - using ! is easily overlooked so I always use '== false'.  So I always add '== true' for symmetry